wcfsetup/install/files/lib/system/flood/FloodControl.class.php

```diff
@@ -0,0 +1,241 @@
+ * @license     GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package     WoltLabSuite\Core\System\Flood
+ * @since       5.4
+ */
+class FloodControl extends SingletonFactory {
+	const PRUNE_TIME = TIME_NOW - 31 * 86400;
+	
+	/**
+	 * Returns the number of contents by a certain identifier type within a certain
+	 * period of time `[$time-$interval, $time]` and the earliest time within the period content
+	 * was created.
+	 */
+	protected function countContentByIdentifier(string $objectType, string $identifier, \DateInterval $interval, int $time): array {
+		$sql = "SELECT  COUNT(*) AS count, MIN(time) AS earliestTime
+			FROM    wcf" . WCF_N . "_flood_control
+			WHERE   objectTypeID = ?
+			        AND identifier = ?
+			        AND time >= ?
+			        AND time <= ?";
+		$statement = WCF::getDB()->prepareStatement($sql);
+		$statement->execute([
+			$this->getObjectTypeID($objectType),
+			$identifier,
+			DateUtil::getDateTimeByTimestamp($time)->sub($interval)->getTimestamp(),
+			$time,
+		]);
+		
+		return $statement->fetchSingleRow();
+	}
+	
+	/**
+	 * Returns the number of contents the active user created of the given type within a certain
+	 * period of time `[$time-$interval, $time]` and the earliest time within the period content
+	 * was created.
+	 */
+	public function countContent(string $objectType, \DateInterval $interval, int $time = TIME_NOW): array {
+		if (WCF::getUser()->userID) {
+			return $this->countUserContent($objectType, WCF::getUser()->userID, $interval, $time);
+		}
+		else {
+			return $this->countGuestContent($objectType, UserUtil::getIpAddress(), $interval, $time);
+		}
+	}
+	
+	/**
+	 * Returns the number of contents a guest created of the given type within a certain period
+	 * of time `[$time-$interval, $time]` and the earliest time within the period content was
+	 * created.
+	 */
+	public function countGuestContent(string $objectType, string $ipAddress, \DateInterval $interval, int $time = TIME_NOW): array {
+		return $this->countContentByIdentifier(
+			$objectType,
+			$this->getUserIdentifier($objectType, $ipAddress),
+			$interval,
+			$time
+		);
+	}
+	
+	/**
+	 * Returns the number of contents a user created of the given type within a certain period
+	 * of time `[$time-$interval, $time]` and the earliest time within the period content was
+	 * created.
+	 */
+	public function countUserContent(string $objectType, int $userID, \DateInterval $interval, int $time = TIME_NOW): array {
+		return $this->countContentByIdentifier(
+			$objectType,
+			$this->getUserIdentifier($objectType, $userID),
+			$interval,
+			$time
+		);
+	}
+	
+	/**
+	 * Returns the identifier used for a guest with the given ip address for content of the
+	 * given object type.
+	 */
+	protected function getGuestIdentifier(string $objectType, string $ipAddress): string {
+		return \hash_hmac(
+			'md5',
+			'guest:' . $ipAddress,
+			'wcf' . WCF_N. '_flood_log' . $objectType,
```

Comment:
```suggestion
			'wcf1_flood_log:' . $objectType,
```

I removed the `WCF_N` to keep it simple. The purpose of the key is to not allow obvious cross-correlation across multiple tables. Having `WCF_N` != 1 is pretty unlikely, making the code more complex than necessary.

Reply:
At the moment, it is quite easy to search for `wcf1` in the code to find instances where `WCF_N` was not used by accident. Until `WCF_N` is automatically replaced in queries, I would also find it surprising to find a hardcoded `wcf1`.

Reply:
Is `$objectType` the full object type identifier here? Then simply using the `$objectType` should be sufficiently unique. But the following would also make sense to me:

```
static::class.':'.$objectType
```

With an optional prefix of `WCF_UUID` for uniqueness across installations.